### PR TITLE
Use retryable http client in Azure authz provider

### PR DESCRIPTION
**Pull request by bingosummer**
We see the timeout error in the checkaccess requrests sometimes.

```
Error occured during authorization check. Please retry again. Error: Checkaccess requests have timed out.: context deadline exceeded
```

This PR moves `withRetryableHttpClient` from Azure authn provider to `utils`, and leverage it in Azure authz provider's `RefreshToken` and `CheckAccess`.

### CODE REVIEW

**Review by karataliu — APPROVED** [review]

**Comment by bingosummer** [comment]
@weinong @AzureMarker could you please have a review? Thanks!

File: authz/providers/azure/azure.go [hunk 1]
```diff
@@ -49,7 +50,8 @@ func init() {
 }
 
 type Authorizer struct {
-	rbacClient *rbac.AccessInfo
+	rbacClient           *rbac.AccessInfo
+	HttpClientRetryCount int
```
Comment (weinong) [inline authz/providers/azure/azure.go:54]: should this property be exported (public)?

File: authz/providers/azure/azure.go [hunk 2]
```diff
@@ -120,6 +124,8 @@ func (s Authorizer) Check(ctx context.Context, request *authzv1.SubjectAccessRev
 		return &authzv1.SubjectAccessReviewStatus{Allowed: true, Reason: rbac.AccessAllowedVerdict}, nil
 	}
 
+	ctx = azureutils.WithRetryableHttpClient(ctx, s.HttpClientRetryCount)
```
Comment (weinong) [inline authz/providers/azure/azure.go:127]: 1. why do we save the http client in the context?
2. as far as I can tell, it still uses [AccessInfo.client](https://github.com/kubeguard/guard/blob/master/authz/providers/azure/rbac/rbac.go#L160), no?

Comment (bingosummer) [reply authz/providers/azure/azure.go:127]: I followed the comment of `WithRetryableHttpClient`.   @AzureMarker **Some of the libraries** respects the http client in the context. Does it mean that it's a best-effort?

```
// WithRetryableHttpClient sets the oauth2.HTTPClient key of the context to an
// *http.Client made from makeRetryableHttpClient.
// Some of the libraries we use will take the client out of the context via
// oauth2.HTTPClient and use it, so this way we can add retries to external code.
```

File: authz/providers/azure/azure.go (hunk 1 above)
Comment (bingosummer) [reply authz/providers/azure/azure.go]: fixed

File: authz/providers/azure/azure.go (hunk 2 above)
Comment (AzureMarker) [reply authz/providers/azure/azure.go]: `net/http` is a standard library, it doesn't reference third party code so there's no way it uses that context key.

This key is used in go-oidc:
https://github.com/kubeguard/guard/blob/e0840f9ce43103cd006c07b1cdb4373125956d3e/vendor/github.com/coreos/go-oidc/oidc.go#L60

The other HTTP client usages hard-code to `go.kubeguard.dev/guard/util/httpclient`'s `DefaultHTTPClient`, which is just a normal HTTP client.

Comment (bingosummer) [reply authz/providers/azure/azure.go]: I made this change https://github.com/kubeguard/guard/pull/372/commits/57e588e83569f709f8fd04b22b8dadca66c24080. Could you please review it?

Comment (bingosummer) [reply authz/providers/azure/azure.go]: @AzureMarker gentle ping

File: auth/providers/azure/graph/obo_tokenprovider.go [hunk 3]
```diff
@@ -74,6 +75,9 @@ func (u *oboTokenProvider) Acquire(ctx context.Context, token string) (AuthRespo
 		klog.V(10).Infoln(cmd)
 	}
 
+	if c, ok := ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		u.client = c
```
Comment (AzureMarker) [inline auth/providers/azure/graph/obo_tokenprovider.go:79]: This shouldn't modify the token provider (`u.client`). Store a reference to the HTTP client in a variable and modify it there.

The issue is if this is called twice (first with a context that holds an HTTP client, then with a context that has no HTTP client), the second call will use the HTTP client of the first.

File: authz/providers/azure/rbac/rbac.go [hunk 4]
```diff
@@ -374,6 +375,10 @@ func (a *AccessInfo) CheckAccess(request *authzv1.SubjectAccessReviewSpec) (*aut
 }
 
 func (a *AccessInfo) sendCheckAccessRequest(ctx context.Context, checkAccessURL url.URL, checkAccessBody *CheckAccessRequest, ch chan *authzv1.SubjectAccessReviewStatus) error {
+	if c, ok := ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		a.client = c
```
Comment (AzureMarker) [inline authz/providers/azure/rbac/rbac.go:379]: Same comment here, don't modify `a.client`, store it in a local variable.

File: authz/providers/azure/azure.go (hunk 2 above)
Comment (AzureMarker) [reply authz/providers/azure/azure.go]: Added some comments

File: auth/providers/azure/graph/obo_tokenprovider.go (hunk 3 above)
Comment (bingosummer) [reply auth/providers/azure/graph/obo_tokenprovider.go]: removed this part to make the pr focus on authz.

File: authz/providers/azure/rbac/rbac.go (hunk 4 above)
Comment (bingosummer) [reply authz/providers/azure/rbac/rbac.go]: fixed. Could you please check whether the fix meets the requirement? Thanks

**Comment by bingosummer** [comment]
@AzureMarker  Could you please have a review again? Thanks

**Review by AzureMarker — APPROVED** [review]

**Comment by bingosummer** [comment]
@weinong could you please have a review? Thanks

**Comment by bingosummer** [comment]
The PR is rebased. @weinong could you please have a review? Please let me know if anything is missing. Thanks!

File: authz/providers/azure/rbac/rbac.go [hunk 5]
```diff
@@ -373,6 +377,11 @@ func (a *AccessInfo) CheckAccess(request *authzv1.SubjectAccessReviewSpec) (*aut
 }
 
 func (a *AccessInfo) sendCheckAccessRequest(ctx context.Context, checkAccessUsername string, checkAccessURL url.URL, checkAccessBody *CheckAccessRequest, ch chan *authzv1.SubjectAccessReviewStatus) error {
+	var clientInCtx *http.Client
+	if c, ok := ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
```
Comment (bcho) [inline authz/providers/azure/rbac/rbac.go:381]: nit: can we move this client loader to `util/azure` so we can hide the context key from caller?

File: authz/providers/azure/azure_test.go [hunk 6]
```diff
@@ -37,7 +37,8 @@ import (
 )
 
 const (
-	loginResp = `{ "token_type": "Bearer", "expires_in": 8459, "access_token": "%v"}`
+	loginResp            = `{ "token_type": "Bearer", "expires_in": 8459, "access_token": "%v"}`
+	httpClientRetryCount = 2
```
Comment (bcho) [inline authz/providers/azure/azure_test.go:41]: can we add a unit test for covering the retry behavior?

File: authz/providers/azure/azure_test.go [hunk 7]
```diff
@@ -52,9 +53,10 @@ func clientSetup(serverUrl, mode string) (*Authorizer, error) {
 	}
 
 	authOpts := auth.Options{
-		ClientID:     "client_id",
-		ClientSecret: "client_secret",
-		TenantID:     "tenant_id",
+		ClientID:             "client_id",
+		ClientSecret:         "client_secret",
+		TenantID:             "tenant_id",
+		HttpClientRetryCount: httpClientRetryCount,
```
Comment (julienstroheker) [inline authz/providers/azure/azure_test.go:59]: Can we assert that it try 2 times ?

File: authz/providers/azure/azure_test.go [hunk 8]
```diff
@@ -171,6 +174,7 @@ func TestCheck(t *testing.T) {
 		assert.Nilf(t, resp, "response should be nil")
 		assert.NotNilf(t, err, "should get error")
 		assert.Contains(t, err.Error(), "Error occured during authorization check")
+		assert.Contains(t, err.Error(), fmt.Sprintf("giving up after %d attempt", httpClientRetryCount+1))
```
Comment (bingosummer) [inline authz/providers/azure/azure_test.go:177]: @bcho @julienstroheker this line can assert that it gives up after 3 attempts.

File: authz/providers/azure/azure_test.go (hunk 6 above)
Comment (bingosummer) [reply authz/providers/azure/azure_test.go:42]: added

File: authz/providers/azure/azure_test.go (hunk 8 above)
Comment (bingosummer) [reply authz/providers/azure/azure_test.go:177]: could you please have another round of review? Thanks

**Review by bcho — APPROVED** [review]
LGTM

**Comment by bingosummer** [comment]
@weinong could you please have another round of review? Thanks

**Comment by weinong** [comment]
@julienstroheker PTAL

**Review by weinong — APPROVED** [review]
approved

**Pull request merged**